### PR TITLE
Support arbitrary output dtypes in PyT GEMM functions

### PR DESCRIPTION
**Pull request by timmoon10**
We currently implement wgrad fusion by assuming that weight tensors contain an FP32 `main_grad` tensor. However, we would like to generalize and support BF16 `main_grad`s (see https://github.com/NVIDIA/NeMo/pull/5920). This PR deprecates the `fp32_output` option in the PyTorch `gemm`/`fp8_gemm` functions and automatically detects the dtype based on the user-provided output tensor. If no output tensor is provided, there is no change in functionality.

Pinging @mikolajblaz and @anmolgupt.

### CODE REVIEW

File: transformer_engine/pytorch/cpp_extensions.py [hunk 1]
```diff
@@ -26,9 +26,9 @@ def fp8_gemm(
     fp8_meta_tensor: tex.FP8TensorMeta = None,
     bias: Optional[torch.Tensor] = None,
     use_bias: bool = False,
-    fp32_output: bool = False,
+    fp32_output: bool = False, # Deprecated
```
Comment (ptrendx) [inline transformer_engine/pytorch/cpp_extensions.py:29]: It is not a public API, we should be able to change that without deprecation.

File: transformer_engine/pytorch/cpp_extensions.py [hunk 2]
```diff
@@ -94,7 +93,7 @@ def gemm(
     out: Optional[torch.Tensor] = None,
     bias: Optional[torch.Tensor] = None,
     use_bias: bool = False,
-    fp32_output: bool = False,
+    fp32_output: bool = False, # Deprecated
```
Comment (ptrendx) [inline transformer_engine/pytorch/cpp_extensions.py:96]: Same.

**Comment by timmoon10** [comment]
The test failure is unrelated and shows up in the `main` branch. See https://github.com/NVIDIA/TransformerEngine/pull/67#issuecomment-1430626755.

**Review by ksivaman — APPROVED** [review]
LGTM, thank you!

**Pull request merged**